File: backend/app/models/scenario.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


Tone = Literal["danger", "warning", "ok", "neutral"]
MetricKey = Literal["containment", "service", "trust", "fatigue"]
ChoiceRank = Literal["recommended", "bad", "neutral"]


class Evidence(BaseModel):
    model_config = ConfigDict(extra="forbid")

    label: str
    value: str
    tone: Tone


class ChoiceEffect(BaseModel):
    model_config = ConfigDict(extra="forbid")

    metrics: dict[MetricKey, int] = Field(default_factory=dict)
    nextNode: str
    tag: str
    rationale: str
    rank: ChoiceRank


class Choice(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    label: str
    description: str
    effect: ChoiceEffect


class ScenarioNode(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    phase: str
    eyebrow: str
    title: str
    body: list[str]
    alert: str
    evidence: list[Evidence]
    choices: list[Choice]


class Ending(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    type: Literal["true", "normal", "bad"]
    title: str
    summary: str


class Scenario(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    subtitle: str
    organization: str
    objective: str
    startNode: str
    recommendedRoute: str | None = None
    nodes: dict[str, ScenarioNode]
    endings: dict[str, Ending]


class ScenarioSummary(BaseModel):
    """一覧 API 用。詳細ノードは含めない。"""

    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    subtitle: str
    organization: str
    objective: str

File: frontend/src/types/scenario.ts
export type MetricKey = 'containment' | 'service' | 'trust' | 'fatigue';
export type Metrics = Record<MetricKey, number>;
export type ChoiceRank = 'recommended' | 'bad' | 'neutral';
export type EndingType = 'true' | 'normal' | 'bad';
export interface Evidence {label:string;value:string;tone:'danger'|'warning'|'ok'|'neutral'}
export interface Choice {id:string;label:string;description:string;effect:{rank:ChoiceRank;nextNode:string;tag:string;rationale:string;metrics:Partial<Metrics>}}
export interface ScenarioNode {id:string;phase:string;eyebrow:string;title:string;body:string[];alert:string;evidence:Evidence[];choices:Choice[]}
export interface Ending {id:string;type:EndingType;title:string;summary:string}
export interface ScenarioSummary {id:string;title:string;subtitle:string;organization:string;objective?:string}
export interface Scenario extends ScenarioSummary {objective:string;startNode:string;recommendedRoute?:string|string[];nodes:Record<string,ScenarioNode>;endings:Record<string,Ending>}
export interface HistoryEntry {nodeId:string;choice:Choice;metrics:Metrics}
export interface GameState {nodeId:string;metrics:Metrics;history:HistoryEntry[];ending:Ending|null}
